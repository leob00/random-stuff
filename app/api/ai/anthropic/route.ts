import Anthropic from '@anthropic-ai/sdk'
import { getUserSSRAppRouteApi } from 'app/serverActions/auth/user'
import { NextRequest } from 'next/server'

const anthropic = new Anthropic({
  apiKey: process.env.ANTHROPIC_API_KEY,
})

export async function POST(request: NextRequest) {
  const user = await getUserSSRAppRouteApi()
  if (!user) {
    return new Response(JSON.stringify({ error: 'Unauthorized. Please sign in.' }), { status: 403 })
  }
  //   if (user && user.roles && !!user.roles.find((m) => m.Name == 'Admin')) {
  //     return new Response(JSON.stringify({ error: 'This feature is only available to certain users.' }), { status: 403 })
  //   }
  try {
    const { message, model = 'claude-3-5-sonnet-20241022' } = await request.json()

    if (!message) {
      return new Response(JSON.stringify({ error: 'Message is required' }), { status: 400, headers: { 'Content-Type': 'application/json' } })
    }

    // Create a readable stream
    const stream = new ReadableStream({
      async start(controller) {
        try {
          const messageStream = await anthropic.messages.create({
            model: model,
            max_tokens: 4096,
            messages: [
              {
                role: 'user',
                content: message,
              },
            ],
            stream: true,
          })

          // Process the stream
          for await (const chunk of messageStream) {
            if (chunk.type === 'content_block_delta') {
              const text = chunk.delta
              if (text) {
                // Send the text chunk to the client
                controller.enqueue(new TextEncoder().encode(`data: ${JSON.stringify({ text })}\n\n`))
              }
            }
            if (chunk.type === 'message_delta') {
              const usage = chunk.usage
              controller.enqueue(new TextEncoder().encode(`usage: ${JSON.stringify({ usage })}\n\n`))
            }
          }

          // Send completion signal
          controller.enqueue(new TextEncoder().encode(`data: ${JSON.stringify({ done: true })}\n\n`))
          controller.close()
        } catch (error: any) {
          console.error('Streaming error:')
          controller.enqueue(new TextEncoder().encode(`data: ${JSON.stringify({ error: error.message })}\n\n`))
          controller.close()
        }
      },
    })

    return new Response(stream, {
      headers: {
        'Content-Type': 'text/plain; charset=utf-8',
        'Cache-Control': 'no-cache',
        Connection: 'keep-alive',
      },
    })
  } catch (error) {
    console.error('API Error:', error)
    return new Response(JSON.stringify({ error: 'Internal server error' }), { status: 500, headers: { 'Content-Type': 'application/json' } })
  }
}
