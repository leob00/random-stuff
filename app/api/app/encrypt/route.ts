import { weakEncrypt } from 'lib/backend/encryption/useEncryptor'
import { NextRequest, NextResponse } from 'next/server'

export async function POST(request: NextRequest) {
  const enc = await request.json()
  const result = weakEncrypt(enc)

  return NextResponse.json(result)
}
