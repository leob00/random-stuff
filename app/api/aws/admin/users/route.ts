import { getUserSSRAppRouteApi } from 'app/serverActions/auth/user'
import { getUserPool } from 'app/serverActions/aws/cognito/cognito'
import { getItem, searchItems } from 'app/serverActions/aws/dynamo/dynamo'
import { UserProfile } from 'lib/backend/api/aws/models/apiGatewayModels'
import { constructUserProfileKey, constructUserSecretSecondaryKey } from 'lib/backend/api/aws/util'
import { userSecretArraySchema } from 'lib/backend/api/models/zModels'
import { withAuth } from 'lib/backend/api/with-auth'
import { NextRequest, NextResponse } from 'next/server'

const handler = async (req: NextRequest) => {
  const result = await getUserPool()
  console.log(result)

  return NextResponse.json(result)
}

export const POST = withAuth(handler)
