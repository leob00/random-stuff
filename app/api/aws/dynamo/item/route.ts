import { getItem } from 'app/serverActions/aws/dynamo/dynamo'
import { getRandomStuff } from 'lib/backend/api/aws/apiGateway/apiGateway'
import { SignedRequest } from 'lib/backend/csr/nextApiWrapper'
import { weakDecrypt } from 'lib/backend/encryption/useEncryptor'
import { NextRequest, NextResponse } from 'next/server'

export async function POST(request: NextRequest) {
  const enc = (await request.json()) as SignedRequest
  const dec = weakDecrypt(enc.data)
  if (!dec || dec.length === 0) {
    return new Response(JSON.stringify({ message: 'validation failed' }), { status: 403 })
  }
  const result = await getItem(dec)
  // const newHeaders = new Headers(request.headers)
  // newHeaders.set('Cache-Control', 'no-store')
  // NextResponse.next({
  //   request: {
  //     headers: newHeaders,
  //   },
  // })
  return NextResponse.json(result)
}
