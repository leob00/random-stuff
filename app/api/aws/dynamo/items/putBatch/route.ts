import { RandomStuffDynamoItem } from 'app/serverActions/aws/dynamo/dynamo'
import { putItems } from 'app/serverActions/aws/dynamo/dynamoBatch'
import { SignedRequest } from 'lib/backend/csr/nextApiWrapper'
import { weakDecrypt } from 'lib/backend/encryption/useEncryptor'
import { NextRequest, NextResponse } from 'next/server'

export async function POST(request: NextRequest) {
  const enc = (await request.json()) as SignedRequest
  const dec = weakDecrypt(enc.data)
  if (!dec || dec.length === 0) {
    return new Response(JSON.stringify({ message: 'validation failed' }), { status: 403 })
  }
  const items = JSON.parse(dec) as RandomStuffDynamoItem[]
  const result = await putItems(items)
  return NextResponse.json(result)
}
