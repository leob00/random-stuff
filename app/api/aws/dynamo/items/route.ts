import { getItem, searchItems } from 'app/serverActions/aws/dynamo/dynamo'
import { SignedRequest } from 'lib/backend/csr/nextApiWrapper'
import { weakDecrypt } from 'lib/backend/encryption/useEncryptor'
import { NextRequest, NextResponse } from 'next/server'

export async function POST(request: NextRequest) {
  const enc = (await request.json()) as SignedRequest
  const dec = weakDecrypt(enc.data)
  if (!dec || dec.length === 0) {
    return new Response(JSON.stringify({ message: 'validation failed' }), { status: 403 })
  }
  const result = await searchItems(dec)
  return NextResponse.json(result)
}
