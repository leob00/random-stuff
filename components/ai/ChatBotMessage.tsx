import { Box, Typography } from '@mui/material'
import CopyableText from 'components/Atoms/Text/CopyableText'
import { useChatbotColors } from './aihelper'
import dayjs from 'dayjs'
import { AnthropicChatbotMessage } from 'app/api/ai/anthropic/route'

const ChatBotMessage = ({ msg }: { msg: AnthropicChatbotMessage }) => {
  const { getColor } = useChatbotColors()
  return (
    <Box>
      <Box display={'flex'} alignItems={'center'} gap={2} py={1} color={getColor(msg.role)}>
        <Typography variant='caption'>{msg.role === 'user' ? 'You: ' : msg.role === 'error' ? 'Error' : 'AI: '}</Typography>
        <Typography variant='caption'>{dayjs(msg.timestamp).format('MM/DD/YYYY hh:mm A')}</Typography>
      </Box>
      <Box>
        {msg.content.includes('```') ? (
          <pre>
            <code>{<CopyableText label='' value={msg.content.replaceAll('```', '')} showValue />}</code>
          </pre>
        ) : (
          <Box>
            <CopyableText label='' value={msg.content} showValue labelColor={getColor(msg.role)} />
          </Box>
        )}
      </Box>
    </Box>
  )
}

export default ChatBotMessage
