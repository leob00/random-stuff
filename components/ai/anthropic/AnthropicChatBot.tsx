import { Anthropic } from '@anthropic-ai/sdk/client'
import { Box, TextField, Typography } from '@mui/material'
import ScrollIntoView from 'components/Atoms/Boxes/ScrollIntoView'
import DangerButton from 'components/Atoms/Buttons/DangerButton'
import SuccessButton from 'components/Atoms/Buttons/SuccessButton'
import CenterStack from 'components/Atoms/CenterStack'
import ComponentLoader from 'components/Atoms/Loaders/ComponentLoader'
import ReadOnlyField from 'components/Atoms/Text/ReadOnlyField'
import dayjs from 'dayjs'
import numeral from 'numeral'
import { useState, useRef } from 'react'
import { CasinoGrayTransparent } from 'components/themes/mainTheme'
import { AnthropicChatbotMessage } from 'app/api/ai/anthropic/route'
import ChatBotMessage from 'components/ai//ChatBotMessage'

const AnthropicChatBot = () => {
  const [messages, setMessages] = useState<AnthropicChatbotMessage[]>([])
  const [input, setInput] = useState('')
  const [isLoading, setIsLoading] = useState(false)
  const [currentResponseText, setCurrentResponseText] = useState('')
  const abortControllerRef = useRef<AbortController | null>(null)

  const streamChatResponse = async (message: string) => {
    const userMessage: AnthropicChatbotMessage = { role: 'user', content: message, timestamp: dayjs().format() }
    setMessages((prev) => [...prev, userMessage])
    setIsLoading(true)

    // Create abort controller for cancellation
    abortControllerRef.current = new AbortController()

    try {
      const response = await fetch('/api/ai/anthropic', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ userMessage }),
        signal: abortControllerRef.current.signal,
      })

      if (!response.ok) {
        const text = await response.text()
        setIsLoading(false)
        setMessages((prev) => [
          ...prev,
          {
            content: JSON.parse(text).error,
            role: 'error',
            timestamp: dayjs().format(),
          },
        ])
        return
      }

      const reader = response.body?.getReader()
      if (reader) {
        const decoder = new TextDecoder()
        let accumulatedResponse = ''
        let responseMessages: AnthropicChatbotMessage[] = []
        let usage: Anthropic.Messages.MessageDeltaUsage | undefined = undefined

        while (true) {
          const { done, value } = await reader.read()
          if (done) {
            break
          }

          const chunk = decoder.decode(value)
          const lines = chunk.split('\n')
          for (const line of lines) {
            if (line.startsWith('usage: ')) {
              const usageData = JSON.parse(line.slice(7))
              if (usageData.usage) {
                if (!usage) {
                  usage = usageData.usage
                }
              }
            }
            if (line.startsWith('data: ')) {
              try {
                const data = JSON.parse(line.slice(6))
                if (data.text) {
                  accumulatedResponse = `${accumulatedResponse}${data.text.text}`
                  setCurrentResponseText(accumulatedResponse)
                } else if (data.done) {
                  // Stream completed - add final message
                  responseMessages.push({
                    role: 'assistant',
                    content: accumulatedResponse,
                    timestamp: dayjs().format(),
                    usage: usage,
                  })
                  setMessages((prev) => [...prev, ...responseMessages])
                  setIsLoading(false)
                  return
                } else if (data.error) {
                  throw new Error(data.error)
                }
              } catch (parseError) {
                console.error('Failed to parse streaming data:', parseError)
              }
            }
          }
        }
      }
    } catch (error: any) {
      if (error.name === 'AbortError') {
        console.log('Request was cancelled')
      } else {
        console.error('Streaming error:', error)
        const errorMessage: AnthropicChatbotMessage = {
          role: 'error',
          content: `Error: ${error.message}`,
          timestamp: dayjs().format(),
        }
        setMessages((prev) => [...prev, errorMessage])
      }
    } finally {
      setCurrentResponseText('')
      setIsLoading(false)
    }
  }

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    if (!input.trim() || isLoading) return

    const messageToSend = input.trim()
    setInput('')
    setCurrentResponseText('')
    await streamChatResponse(messageToSend)
  }

  const handleStop = () => {
    if (abortControllerRef.current) {
      abortControllerRef.current.abort()
    }
  }

  return (
    <>
      <Box display={'flex'} flexDirection={'column'} gap={2}>
        <Box>
          {messages.map((msg, index) => (
            <Box key={index}>
              {msg.content && <ChatBotMessage msg={msg} />}
              {msg.usage && (
                <Box display={'flex'} flexDirection={'row'} gap={2}>
                  <ReadOnlyField
                    color={CasinoGrayTransparent}
                    variant='caption'
                    label='input tokens'
                    val={`${numeral(msg.usage.input_tokens).format('###,###')}`}
                  />
                  <ReadOnlyField
                    color={CasinoGrayTransparent}
                    variant='caption'
                    label='output tokens'
                    val={`${numeral(msg.usage.output_tokens).format('###,###')}`}
                  />
                </Box>
              )}
            </Box>
          ))}
        </Box>

        <Box>
          <form onSubmit={handleSubmit}>
            {!isLoading ? (
              <TextField
                placeholder='ask me anything...'
                multiline
                rows={4}
                sx={{ width: '100%' }}
                value={input}
                onChange={(e) => setInput(e.target.value)}
                slotProps={{
                  input: {
                    autoCorrect: 'off',
                  },
                }}
              />
            ) : (
              <Box>
                <ScrollIntoView />
                <Box minHeight={100}>{currentResponseText && <Typography>{currentResponseText}</Typography>}</Box>
              </Box>
            )}

            <Box>
              {isLoading ? (
                <Box>
                  <Box display={'flex'} justifyContent={'center'}>
                    <Box display={'flex'} flexDirection={'column'} gap={1}>
                      <ComponentLoader pt={14} />
                      <CenterStack>
                        <Typography py={8}>thinking...</Typography>
                      </CenterStack>
                      <Box>
                        <DangerButton text='stop' type='button' onClick={handleStop}>
                          Stop
                        </DangerButton>
                      </Box>
                    </Box>
                  </Box>
                </Box>
              ) : (
                <Box py={2} display={'flex'} justifyContent={'flex-end'}>
                  <SuccessButton size='small' text='Go' type='submit' loading={isLoading} disabled={!input.trim()}>
                    Go
                  </SuccessButton>
                </Box>
              )}
            </Box>
          </form>
        </Box>
      </Box>
    </>
  )
}

export default AnthropicChatBot
