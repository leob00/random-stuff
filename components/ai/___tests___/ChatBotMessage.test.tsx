import { render } from '@testing-library/react'
import { AnthropicChatbotMessage } from 'app/api/ai/anthropic/route'
import React from 'react'
import dayjs from 'dayjs'
import ChatBotMessage from '../ChatBotMessage'
describe('AnthropicChatbotMessage component', () => {
  it('matches the snapshot', () => {
    const msg: AnthropicChatbotMessage = {
      role: 'user',
      content: 'hello, how are you?',
      timestamp: dayjs(new Date(2025, 7, 1)).format(),
      usage: {
        input_tokens: 10,
        output_tokens: 100,
        cache_creation_input_tokens: 0,
        cache_read_input_tokens: 0,
        server_tool_use: null,
      },
    }
    const { container } = render(<ChatBotMessage msg={msg} />)
    expect(container).toMatchSnapshot()
  })
})
