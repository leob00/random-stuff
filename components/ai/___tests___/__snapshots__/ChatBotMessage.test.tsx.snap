// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AnthropicChatbotMessage component matches the snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-t3yh2f"
    >
      <span
        class="MuiTypography-root MuiTypography-caption css-mpaqru-MuiTypography-root"
      >
        You: 
      </span>
      <span
        class="MuiTypography-root MuiTypography-caption css-mpaqru-MuiTypography-root"
      >
        08/01/2025 12:00 AM
      </span>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiBox-root css-1yp4ln"
        >
          <div
            class="MuiBox-root css-0"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-13f3v65-MuiTypography-root"
            >
              hello, how are you?
            </p>
          </div>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-155wbr6-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeSmall css-os5i4q-MuiSvgIcon-root"
              data-testid="ContentCopyIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2m0 16H8V7h11z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
